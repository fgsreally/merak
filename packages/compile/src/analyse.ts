import { resolve } from 'path'
import { parseSync, traverse } from '@babel/core'
import type { IText } from 'html5parser'
import { parse, walk } from 'html5parser'
import MagicString from 'magic-string'
import type { MerakAttrs, MerakHTMLFile, MerakJSFile } from './types'
import { desctructGlobal, isCdn, relativePath } from './utils'
export const analyseHTML = (code: string) => {
  const ast = parse(code)
  const ret = {
    _f: '',
    _s: [],
    _l: {},
  } as unknown as MerakHTMLFile

  walk(ast, {
    enter: (node) => {
      if (node.type === 'Tag') {
        if (node.name === 'script') {
          const outScriptNode = node.attributes.find(item => item.name.value === 'src')
          const merakAttrs = {} as unknown as MerakAttrs

          // node.attributes.filter(item => item.name.value.startsWith('data-merak-')).forEach((item) => {
          //   merakAttrs[item.name.value.slice(11)] = item.value?.value || true
          // })
          if (node.attributes.some(item => item.name.value === 'merak-ignore'))
            return
          node.attributes.filter(item => item.name.value !== 'src' && item.name.value !== 'merak-src').forEach((item) => {
            merakAttrs[item.name.value] = item.value?.value || true
          })
          // recording external script node to replace them in browser
          if (outScriptNode) {
            const merakSrc = node.attributes.find(item => item.name.value === 'merak-src')?.value?.value

            const { start, end } = node
            const { value: { value } } = outScriptNode as any
            ret._s.push({ _f: merakSrc || value, _tl: [start, end], _a: merakAttrs, _t: 'outline' })
          }
          else {
            const scriptType = node.attributes.find(item => item.name.value === 'type')?.value?.value
            const { start: tagStart, end: tagEnd } = node
            const { start } = node.close as IText
            const { end } = node.open

            const source = code.slice(end, start)

            const { _l } = analyseInlineESM(source)
            switch (scriptType) {
              case undefined:

                ret._s.push({ _tl: [tagStart, tagEnd], _b: [end, start], _t: 'iife', _a: merakAttrs, _l })
                break
              case 'module':

                ret._s.push({ _tl: [tagStart, tagEnd], _b: [end, start], _t: 'esm', _a: merakAttrs, _l })
            }
          }
        }

        // recording link node to replace publicpath in browser(useless in pure mode)

        if (node.name === 'link' && node.attributes.some(item => item.name.value === 'href')) {
          const outLinkNode = node.attributes.find(item => item.name.value === 'href')
          const { value: { start, end, value } } = outLinkNode as any
          ret._l[value] = { _l: [start, end] }
        }
      }
    },
  })
  return ret
}

export const analyseJSGlobals = (code: string, globals: string[]) => {
  const globalSet = new Set<string>()
  const globalUsed = new Set<string>()
  const ast = parseSync(code)
  traverse(ast, {
    // record all global variable
    Identifier(path) {
      const { scope } = path
      Object.keys((scope as any).globals).forEach((item) => {
        globalSet.add(item)
        globals.includes(item) && globalUsed.add(item)
      })
    },
    // static import
    //  relative path
  })

  return {
    globals: [...globalSet],
    globalUsed: [...globalUsed],
  }
}

export function analyseInlineESM(code: string) {
  const importLoc = [] as [number, number][]
  const ast = parseSync(code)

  traverse(ast, {
    // record all global variable
    // Identifier(path) {
    //   const { scope } = path
    //   Object.keys((scope as any).globals).forEach(item => globals.includes(item) && globalSet.add(item))
    // },
    // static import
    //  relative path
    ImportDeclaration(path) {
      const { node } = path

      const { value, start, end } = node.source as any
      if (value.startsWith('.') || value.startsWith('/'))
        importLoc.push([start, end])
    },
    // dynamic import
    Import(path) {
      const { start, end } = path.parent as { start: number; end: number }
      const { value } = (path.parent as any).arguments[0]
      if (value.startsWith('.') || value.startsWith('/'))
        importLoc.push([start, end])
    },

  })
  return { _l: importLoc }
}

/**
 * @deprecated
 */
export const analyseJS = (code: string, filePath: string, rootPath: string, globals: string[]) => {
  const ret = {
    type: 'js',
    filePath: relativePath(rootPath, filePath),
    imports: [],
    dynamicImports: {},
  } as unknown as MerakJSFile
  const ast = parseSync(code)

  const globalSet = new Set()
  traverse(ast, {
    // record all global variable
    Identifier(path) {
      const { scope } = path
      Object.keys((scope as any).globals).forEach(item => globals.includes(item) && globalSet.add(item))
    },
    // static import
    //  relative path
    ImportDeclaration(path) {
      const { node } = path

      const { value, start, end } = node.source as any
      if (isCdn(value))
        ret.imports.push({ filePath: '', loc: [start, end] })

      else
        ret.imports.push({ filePath: relativePath(rootPath, resolve(filePath, '../', value)), loc: [start, end] })
    },
    // dynamic import
    Import(path) {
      const { start, end } = path.parent as { start: number; end: number }
      const { value } = (path.parent as any).arguments[0]
      ret.dynamicImports[relativePath(rootPath, resolve(filePath, '../', value))] = { loc: [start, end] }
    },

  })
  ret.globals = [...globalSet] as string[]
  return ret
}

export function injectGlobalToIIFE(code: string, globalVar: string, globals: string[]) {
  const globalSet = new Set<string>()
  const ast = parseSync(code, { filename: 'any' })
  const s = new MagicString(code)

  traverse(ast, {

    Identifier(path) {
      const { scope } = path
      Object.keys((scope as any).globals).forEach(item => globals.includes(item) && globalSet.add(item))
    },
    Program(path) {
      const { start, end } = path.node
      s.appendLeft(start || 0, `(()=>{const {${desctructGlobal([...globalSet])}}=${globalVar};`)
      s.appendRight(end || 0, '})()')
    },
  })
  return { code: s.toString(), map: s.generateMap({ hires: true }) }
}

export function injectGlobalToESM(code: string, globalVar: string, globals: string[]) {
  const globalSet = new Set<string>()
  const ast = parseSync(code)
  const s = new MagicString(code)
  let lastImport
  traverse(ast, {

    Identifier(path) {
      const { scope } = path
      Object.keys((scope as any).globals).forEach(item => globals.includes(item) && globalSet.add(item))
    },
    ImportDeclaration(path) {
      const { node } = path

      const { end } = node.source
      lastImport = end
    },
  })

  s.appendRight(lastImport || 0, `\nconst {${desctructGlobal([...globalSet])}}=${globalVar};`)

  return { code: s.toString(), map: s.generateMap({ hires: true }) }
}
