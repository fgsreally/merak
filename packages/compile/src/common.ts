export const DEFAULT_NATIVE_VARS = ['document', 'history', 'location', 'window', 'self', 'globalThis', 'setTimeout', 'setInterval', 'addEventListener']

export const EXCLUDE_TAG = '/** @merak */'
