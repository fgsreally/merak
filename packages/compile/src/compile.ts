/**
 * this file was working for dynamic mode,to inject globalVars in browser
 * but now, merak perfers to inject globalVars by compiling, the file is  useless
 * I plan to provide an new dynamic mode in future,
 * so keep the file
 */
import { extname, join, relative, resolve } from 'path'
import fs from 'fs-extra'
import {
  analyseHTML,
  analyseJS,
  injectGlobalToESM,
  injectGlobalToIIFE,
} from './analyse'

import { isCdn } from './utils'
import type { ImportScript, MerakConfig, MerakJSFile, MerakOptions, SpaMerakConfig } from './types'
// const fileSet = new Set()

/**
 * @deprecated
 */
export async function compileSsrProject(inputOptions: any, root: string) {
  const allInfo = {}
  for (const filePath of inputOptions.files) {
    const jsInfo = { files: [], lazyFiles: {} } as SpaMerakConfig
    await compileJS(inputOptions, jsInfo, root, filePath, new Set())

    allInfo[filePath] = jsInfo
    for (const fileInfo of jsInfo.files) {
      for (const dynamicPath in (fileInfo as MerakJSFile).dynamicImports) {
        const dynamicConfig = jsInfo.lazyFiles[dynamicPath] = { files: [], lazyFiles: [] }
        await compileLazyJS(inputOptions, dynamicConfig, jsInfo, root, dynamicPath, new Set())
      }
    }
  }
  return allInfo
}

export async function compileSpaProject(inputOptions: MerakOptions, root: string) {
  const projectInfo = { files: [], lazyFiles: {} } as SpaMerakConfig

  const htmlPath = resolve(root, 'index.html')
  const str = await fs.promises.readFile(htmlPath, 'utf-8')
  const htmlInfo = analyseHTML(str)

  for (const importInfo of htmlInfo.scripts) {
    if ((importInfo as ImportScript).filePath)
      await compileJS(inputOptions, projectInfo, root, (importInfo as ImportScript).filePath, new Set())
  }

  for (const fileInfo of projectInfo.files) {
    for (const dynamicPath in (fileInfo as MerakJSFile).dynamicImports) {
      const dynamicConfig = projectInfo.lazyFiles[dynamicPath] = { files: [], lazyFiles: [] } as any
      await compileLazyJS(inputOptions, dynamicConfig, projectInfo, root, dynamicPath, new Set())
    }
  }
  projectInfo.files.push(htmlInfo)

  return projectInfo
}

/**
 * @deprecated
 */
export async function compileJS(inputOptions: MerakOptions, config: SpaMerakConfig, root: string, path: string, fileSet: Set<string>) {
  if (config.files.some(item => item.type === 'js' && item.filePath === path))
    return

  if (isCdn(path)) {
    // TODO CDN PATH
  }
  else {
    const jsPath = resolve(root, path)
    if (fileSet.has(jsPath))
      return
    fileSet.add(jsPath)
    const code = await fs.promises.readFile(jsPath, 'utf-8')
    const jsInfo = analyseJS(code, jsPath, root, inputOptions.globals)

    for (const importInfo of jsInfo.imports)

      await compileJS(inputOptions, config, root, importInfo.filePath, fileSet)

    config.files.push(jsInfo)
  }
}

export async function compileLazyJS(inputOptions: MerakOptions, config: {
  files: MerakJSFile[]
  lazyFiles: string[]
}, entryInfo: SpaMerakConfig, root: string, path: string, fileSet: Set<string>) {
  if (config.files.some(item => item.type === 'js' && item.filePath === path))
    return
  if (isCdn(path)) {
    //
  }
  else {
    const jsPath = resolve(root, path)
    if (fileSet.has(jsPath))
      return
    fileSet.add(jsPath)

    const code = await fs.promises.readFile(jsPath, 'utf-8')
    const jsInfo = analyseJS(code, jsPath, root, inputOptions.globals)

    for (const importInfo of jsInfo.imports)
      await compileLazyJS(inputOptions, config, entryInfo, root, importInfo.filePath, fileSet)

    for (const dynamicPath in jsInfo.dynamicImports) {
      const dynamicConfig = entryInfo.lazyFiles[dynamicPath] = { files: [], lazyFiles: [] }
      config.lazyFiles.push(dynamicPath)
      await compileLazyJS(inputOptions, dynamicConfig, entryInfo, root, dynamicPath, fileSet)
    }
    config.files.push(jsInfo)
  }
}
/**
 * @deprecated
 */
export async function compileESMProject(inputOptions: MerakOptions, dir: string, to: string, globalVar: string) {
  const htmlRet = {} as unknown as MerakConfig
  walkSync(dir, (filePath) => {
    if (extname(filePath) === '.js') {
      fs.outputFile(resolve(to, relative(dir, filePath)), injectGlobalToESM(fs.readFileSync(filePath, 'utf-8'), globalVar, inputOptions.globals).code, 'utf-8')
      return
    }
    if (extname(filePath) === '.html')
      htmlRet.template = analyseHTML(fs.readFileSync(filePath, 'utf-8'))

    fs.copySync(filePath, resolve(to, relative(dir, filePath)))
  })
}
/**
 * @deprecated
 */
export async function compileIIFEProject(inputOptions: MerakOptions, dir: string, to: string, globalVar: string) {
  const htmlRet = {} as unknown as MerakConfig
  walkSync(dir, async (filePath) => {
    if (extname(filePath) === '.js') {
      await fs.outputFile(resolve(to, relative(dir, filePath)), injectGlobalToIIFE(fs.readFileSync(filePath, 'utf-8'), globalVar, inputOptions.globals).code, 'utf-8')
      return
    }
    if (extname(filePath) === '.html')
      htmlRet.template = analyseHTML(fs.readFileSync(filePath, 'utf-8'))

    await fs.copy(filePath, resolve(to, relative(dir, filePath)))
  })
  return htmlRet
}
function walkSync(currentDirPath: string, callback: (param: string) => void) {
  fs.readdirSync(currentDirPath, { withFileTypes: true }).forEach((dirent) => {
    const filePath = join(currentDirPath, dirent.name)
    if (dirent.isFile())
      callback(filePath)

    else if (dirent.isDirectory())
      walkSync(filePath, callback)
  })
}
