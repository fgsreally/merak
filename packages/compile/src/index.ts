export * from './analyse'
export * from './types'
export * from './utils'
export * from './postcss'
export * from './common'
export * from './stream'
export * from './log'
