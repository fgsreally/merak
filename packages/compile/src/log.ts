import json2md from 'json2md'
import fse from 'fs-extra'
export class Logger {
  // work for DANGER_IDENTIFIERS
  dangerUsedRecord: Record<string, Record<string, { times: number; loc: [number, number] }>> = {}
  unusedGlobalRecord: Record<string, string[]> = {}

  collectDangerUsed(file: string, message: string, loc: [number, number]) {
    if (!this.dangerUsedRecord[file])
      this.dangerUsedRecord[file] = {}
    if (!this.dangerUsedRecord[file][message])
      this.dangerUsedRecord[file][message] = { times: 0, loc }
    this.dangerUsedRecord[file][message].times++
  }

  collectUnusedGlobals(file: string, globals: string[]) {
    this.unusedGlobalRecord[file] = globals
  }

  output(outputPath?: string) {
    if (!outputPath)
      return
    const ret = [] as any[]
    ret.push({ h1: 'danger used like eval:' })

    for (const file in this.dangerUsedRecord) {
      ret.push({ blockquote: `file:\`${file}\`` })
      for (const message in this.dangerUsedRecord[file]) {
        const { times, loc } = this.dangerUsedRecord[file][message]
        ret.push({ h3: `\`${message}\`` })

        ret.push({ ul: [`loc:\`${file}:${loc[0]}:${loc[1]}\``, `times: \`${times}\``] })
      }
    }

    ret.push({ h1: 'unused globals:' })

    for (const file in this.unusedGlobalRecord) {
      ret.push({ blockquote: `file:\`${file}\`` })
      ret.push({ ol: this.unusedGlobalRecord[file] })
    }
    // eslint-disable-next-line no-console
    console.log(`[MERAK] generate analyse file at ${outputPath}`)
    fse.outputFileSync(outputPath, json2md(ret))
  }
}

export const logger = new Logger()
