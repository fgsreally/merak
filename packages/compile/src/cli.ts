import { resolve } from 'path'
import { createRequire } from 'module'
import fse from 'fs-extra'
import cac from 'cac'
import fg from 'fast-glob'
// @ts-expect-error misstypes
import isVarName from 'is-var-name'
import { analyseHTML, analyseJSGlobals, injectGlobalToESM, injectGlobalToIIFE } from './analyse'
import { DEFAULT_INJECT_GLOBALS } from './common'
import { logger } from './log'
const cli = cac('merak')
const root = process.cwd()
const require = createRequire(root)

cli.command('init', 'create merak.config.json').action(() => {
  fse.outputFile(resolve(root, 'merak.config.json'), JSON.stringify({
    $schema: 'https://unpkg.com/merak-compile/assets/schema.json',
    fakeGlobalVar: '',
  }))
})

cli.command('', 'parse all file to merak-format')
  .option('-c, --config', 'config file', {
    default: 'merak.config.json',
  })

  .action(async (options) => {
    let {
      dir = '', nativeVars = [], customVars = [], fakeGlobalVar, format = 'esm', isinLine = true, outDir = 'dist', includes = ['index.html', '**/*.js', '*.js', '**/*.css'], exclude = ['node_modules/**/*'], logPath,
    } = getConfig(options.config)
    if (!isVarName(fakeGlobalVar))
      throw new Error(`${fakeGlobalVar} is not a valid var`)

    const cwd = resolve(root, dir)
    fse.ensureDir(resolve(root, outDir))
    const entries = await fg(includes, { cwd, ignore: exclude })
    nativeVars = [...new Set([...DEFAULT_INJECT_GLOBALS, ...nativeVars])]
    console.table(entries)
    console.table(nativeVars)
    for (const entry of entries) {
      const filePath = resolve(root, outDir, entry)
      const raw = await fse.readFile(resolve(cwd, entry), 'utf-8')

      if (entry.endsWith('.js')) {
        if (format === 'esm') {
          const { code, warning } = injectGlobalToESM(raw, fakeGlobalVar, nativeVars, customVars)
          warning.forEach(warn => logger.collectDangerUsed(entry, warn.info, [warn.loc.start.line, warn.loc.start.column]))

          await fse.outputFile(filePath, code)
        }
        else {
          const { code, warning } = injectGlobalToIIFE(raw, fakeGlobalVar, nativeVars, customVars)
          warning.forEach(warn => logger.collectDangerUsed(entry, warn.info, [warn.loc.start.line, warn.loc.start.column]))

          await fse.outputFile(filePath, code)
        }
        if (logPath) {
          const unUsedGlobals = analyseJSGlobals(raw, [...nativeVars, ...customVars])
          unUsedGlobals.length > 0 && logger.collectUnusedGlobals(entry, unUsedGlobals)
        }

        return
      }

      if (entry.endsWith('.html')) {
        const merakConfig = {
          _f: fakeGlobalVar, _n: nativeVars, _c: customVars,
        } as any
        let html = raw.replace('</head>', `</head><script merak-ignore>const ${fakeGlobalVar}=window.${fakeGlobalVar}||window</script>`)
        merakConfig._l = analyseHTML(html)
        if (isinLine)
          html = html.replace('</body>', `<m-b config="${JSON.stringify(merakConfig)}"></m-b></body>`)
        else await fse.outputFile(resolve(filePath, '../merak.json'), JSON.stringify(merakConfig), 'utf-8')

        await fse.outputFile(filePath, html, 'utf-8')
        return
      }
      await fse.copyFile(resolve(cwd, entry), filePath)
    }
    logPath && logger.output(resolve(root, logPath))
  })

function getConfig(configPath: string) {
  /**
   * {
   *  "nativeVars":[...],
   * "customVars":[...],
   *  "fakeGlobalVar":'..'
   *  "dir":'..',
   *  "isInline":true/false,
   *  "format":esm/iife
   * }
   */
  return require(resolve(root, configPath))
}

cli.help()
cli.version(require('../package.json').version)

cli.parse()
