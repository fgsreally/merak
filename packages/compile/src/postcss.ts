import { logger } from './log'

export function merakPostCss() {
  return {
    postcssPlugin: 'postcss-root-to-host',

    Root(root: any) {
      root.walkAtRules('font-face', ({ source: { start, input: { file } } }: any) => {
        logger.collectDangerUsed(file, '"@font-face" need manual process', [start.line, start.column])
      })
      root.walk((node: any) => {
        if (node.selector) {
          if (node.selectors.includes(':root') && !node.selectors.includes(':host'))
            node.selector = node.selector.replace(':root', ':root,:host')

          // node.selector = node.selectors
          //   .map((selector: any) => {
          //     if (/:root/g.test(selector) && !/:host/g.test(selector))
          //       return selector.replace(/:root/g, ':root,:host')

          //     return selector
          //   })
          //   .join(', ')
        }
      })
    },
  }
}
