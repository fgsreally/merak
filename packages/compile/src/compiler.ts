import { parseSync, traverse } from /* merak */'@babel/core'
import type { SourceLocation } from '@babel/types'
import type { IText } from 'html5parser'
import { parse, walk } from 'html5parser'
import MagicString from 'magic-string'
import postcss from 'postcss'
import { desctructVars } from './utils'
import { DANGER_IDENTIFIERS, EXCLUDE_TAG } from './common'
import { logger } from './log'
import { merakPostCss } from './postcss'
import { getPathLocFromHTML } from './analyse'

export class Compiler {
  warnings: { info: string; loc: SourceLocation }[] = []
  logger = logger
  unUsedVars = new Set<string>()
  constructor(public projectGlobalVar: string, public nativeVars: string[], public customVars: string[]) {

  }

  get internalScript() {
    return `const ${this.projectGlobalVar}=window.${this.projectGlobalVar}||window;${this.customVars.length > 0 ? `${this.projectGlobalVar}.__m_p__=(k)=>new Proxy(()=>{},{get(_, p) {const v= ${this.projectGlobalVar}[k][p];return typeof v==='function'?v.bind(${this.projectGlobalVar}):v},has(target, p) { return p in ${this.projectGlobalVar}[k]}, set(_,p,v){${this.projectGlobalVar}[k][p]=v;return true },apply(_,t,a){return ${this.projectGlobalVar}[k](...a) }})` : ''}`
  }

  get variables() {
    return [...this.nativeVars, ...this.customVars]
  }

  get initAllVarString() {
    return `const {${desctructVars(this.nativeVars)}}=${this.projectGlobalVar};${this.customVars.map(item => `const ${item}=${this.projectGlobalVar}.__m_p__('${item}')`).reduce((p, c) => `${p};${c}`)}`
  }

  createTag(config: any) {
    return `<merak c='${encodeURIComponent(JSON.stringify({
      n: this.nativeVars,
      c: this.customVars,
      f: this.projectGlobalVar,
      ...config,
    }))}'></merak>`
  }

  compileScript(code: string, file: string, force?: boolean) {
    const nativeSet = new Set<string>()
    const s = new MagicString(code)

    let start = 0
    let end = 0
    if (force) {
      s.appendLeft(start, `(()=>{${this.initAllVarString}`)
      s.append('})()')
    }
    else {
      const ast = parseSync(code, { filename: 'any' }) as any

      traverse(ast, {

        ReferencedIdentifier: (path) => {
          const { start, loc, name } = path.node as any
          if (!path.scope.hasBinding(name, true)) {
            if (!this.variables.includes(name))
              this.logger.collectUnscopedVar(file, name)
            if (DANGER_IDENTIFIERS.includes(path.node.name))
              this.logger.collectDangerUsed(file, `"${name}" is danger,need to be wrapped in $sandbox`, loc)

            if (this.nativeVars.includes(name)) {
              if (s.slice(start! - 13, start!) === EXCLUDE_TAG) {
                s.overwrite(start, start + name.length, `(window.isMerak?window.rawWindow.${name}:${name})`)
                nativeSet.add('window')

                return
              }

              nativeSet.add(name)
            }
            if (this.customVars.includes(name))
              s.appendLeft(start, `${this.projectGlobalVar}.`)
          }
        },
        Program(path) {
          start = path.node.start as number
          end = path.node.end as number
        },
      })
      const injectGlobals = [...nativeSet]

      if (injectGlobals.length) {
        s.appendLeft(start, `(()=>{const {${desctructVars(injectGlobals)}}=${this.projectGlobalVar};`)
        s.appendRight(end, '})()')
      }
    }

    return { code: s.toString(), map: s.generateMap({ hires: true }) }
  }

  compileESM(code: string, file: string, force?: boolean) {
    const nativeSet = new Set<string>()

    const s = new MagicString(code)

    if (force) {
      s.prepend(this.initAllVarString)
    }
    else {
      const ast = parseSync(code) as any
      // let lastImport
      traverse(ast, {

        ReferencedIdentifier: (path) => {
          const { name, loc, start } = path.node
          if (!path.scope.hasBinding(name, true)) {
            if (!this.variables.includes(name))
              this.logger.collectUnscopedVar(file, name)
            if (DANGER_IDENTIFIERS.includes(path.node.name))
              this.logger.collectDangerUsed(file, `"${name}" is danger,need to be wrapped in $sandbox`, [loc!.start.line, loc!.start.column])

            if (this.nativeVars.includes(name)) {
              if (start && s.slice(start! - 13, start!) === EXCLUDE_TAG) {
                s.overwrite(start, start + name.length, `(window.isMerak?window.rawWindow.${name}:${name})`)
                nativeSet.add('window')
                return
              }

              nativeSet.add(name)
            }

            if (this.customVars.includes(name))
              s.appendLeft(start!, `${this.projectGlobalVar}.`)
          }
        },

      })

      const injectGlobals = [...nativeSet]
      // globals = injectGlobals
      if (injectGlobals.length)
        s.appendRight(0, `\nconst {${desctructVars(injectGlobals)}}=${this.projectGlobalVar};`)
    }
    return { code: s.toString(), map: s.generateMap({ hires: true }) }
  }

  compileStatement(code: string) {
    const s = new MagicString(code)
    const ast = parseSync(code, { filename: 'any' }) as any
    traverse(ast, {
      ReferencedIdentifier: (path) => {
        const name = path.node.name
        if (!path.scope.hasBinding(name, true)) {
          const { start } = path.node as { start: number }
          s.appendRight(start, `${this.projectGlobalVar}.`)
        }
      },
    })

    return { code: s.toString(), map: s.generateMap({ hires: true }) }
  }

  compileStyle(style: string, file: string) {
    const { css, map } = postcss([merakPostCss()]).process(style, { from: file })
    return { code: css, map }
  }

  compileHTML(html: string, file: string) {
    const s = new MagicString(html)
    s.replace('</head>', `<head><script merak-ignore>${this.internalScript}</script>`)
    const ast = parse(html)

    walk(ast, {
      enter: (node) => {
        if (node.type === 'Tag') {
          if (node.name === 'script') {
            const outScriptNode = node.attributes.some(item => item.name.value === 'src')
            const isModule = node.attributes.some(item => item.name.value === 'type' && item.value?.value === 'module')
            if (node.attributes.some(item => item.name.value === 'merak-ignore'))
              return

            // recording external script node to replace them in browser
            if (!outScriptNode) {
              const { start } = node.close as IText
              const { end } = node.open
              const source = html.slice(end, start)

              const { code } = isModule ? this.compileESM(source, file) : this.compileScript(source, file)

              s.overwrite(end, start, code)
            }
          }

          if (node.name === 'style') {
            const { start } = node.close as IText
            const { end } = node.open
            const source = html.slice(end, start)
            const { code } = this.compileStyle(source, file)
            s.overwrite(end, start, code)
          }

          if (node.attributeMap) {
            Object.values(node.attributeMap).forEach(({ name, value }) => {
              if (name.value.startsWith('on') && value)

                s.overwrite(value.start, value.end, this.compileStatement(value.value).code)
            })
          }
        }
      },
    })

    const loc = getPathLocFromHTML(s.toString())

    s.replace('</body>', `${this.createTag({ l: loc })}</body>`)

    return { code: s.toString(), map: s.generateMap({ hires: true }) }
  }

  output(outputPath: string) {
    this.logger.output(outputPath)
  }
}
