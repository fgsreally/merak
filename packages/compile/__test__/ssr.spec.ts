/* eslint-disable promise/param-names */
import fs from 'fs'
import { resolve } from 'path'
import { Writable } from 'node:stream'
import { describe, expect, it } from 'vitest'
import { Compiler, SsrTransformer, mergeCompiledHTML } from '../src'
import html from './fixtures/index.html?raw'

describe('ssr', () => {
  const app1 = 'https://localhost:3000/app1/index.html'
  const app2 = 'https://localhost:3000/app2/index.html'

  const compiler1 = new Compiler('app1', ['window', 'document'], [])
  const compiler2 = new Compiler('app2', ['window', 'document'], [])

  it('basic', async () => {
    expect(mergeCompiledHTML(compiler1.compileHTML(html, 'index.html').code, compiler2.compileHTML(html, 'index.html').code, app2, { attrs: { class: 'test' } }),
    ).toMatchSnapshot()
  })

  it('stream', () => new Promise<void>((done) => {
    const mainStream = fs.createReadStream(resolve(__dirname, './fixtures/index.html'))
    const subStream = fs.createReadStream(resolve(__dirname, './fixtures/index.html'))

    const transfomer = new SsrTransformer(app1)

    const mockHttpStream = new Writable()

    let data = ''
    mockHttpStream._write = (chunk, encoding, callback) => {
      data += chunk.toString()
      callback()
    }

    transfomer.on('end', () => {
      expect(data).toMatchSnapshot()
      done()
    })

    mainStream.on('end', () => {
      transfomer.pipe(mockHttpStream)
    })

    mainStream.pipe(mockHttpStream, { end: false })

    subStream.pipe(transfomer)
  }))
})
