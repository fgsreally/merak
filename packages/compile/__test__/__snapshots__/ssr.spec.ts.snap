// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ssr > basic 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <meta charset=\\"UTF-8\\" />
    <link rel=\\"icon\\" href=\\"/favicon.ico\\" />
    <link rel=\\"stylesheet\\" href=\\"/assets/index-28fadf32.css\\" />
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
  </head>

  <body>
    <script type=\\"module\\" src=\\"/src/main.js\\"></script>
    <script type=\\"module\\">
      import y from 'y'
import xx from './xx'
document.write(xx)
console.log(y)
    </script>
    <script>
      window.addEventListener('click', async () => {
        await import('./esm.js')
})
    </script>
  <template data-merak-url='https://localhost:3000/app/index.html' class='test'><!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <meta charset=\\"UTF-8\\" />
    <link rel=\\"icon\\" href=\\"https://localhost:3000/favicon.ico\\" />
    <link rel=\\"stylesheet\\" href=\\"https://localhost:3000/assets/index-28fadf32.css\\" />
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
  </head>

  <body>
    <script type=\\"module\\" src=\\"https://localhost:3000/src/main.js\\"></script>
    <script type=\\"module\\">
      import y from 'y'
import xx from 'https://localhost:3000/app/xx'
document.write(xx)
console.log(y)
    </script>
    <script>
      window.addEventListener('click', async () => {
        await import('https://localhost:3000/app/esm.js')
})
    </script>
  </body>
</html>
</template></body>
</html>
"
`;

exports[`ssr > stream 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <meta charset=\\"UTF-8\\" />
    <link rel=\\"icon\\" href=\\"/favicon.ico\\" />
    <link rel=\\"stylesheet\\" href=\\"/assets/index-28fadf32.css\\" />
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
  </head>

  <body>
    <script type=\\"module\\" src=\\"/src/main.js\\"></script>
    <script type=\\"module\\">
      import y from 'y'
import xx from './xx'
document.write(xx)
console.log(y)
    </script>
    <script>
      window.addEventListener('click', async () => {
        await import('./esm.js')
})
    </script>
  </body>
</html>
<template data-merak-url='https://localhost:3000/app/index.html' >
<html lang='en'>
  <head>
    <meta charset='UTF-8'></meta>
    <link rel='icon' href='https://localhost:3000/favicon.ico'></link>
    <link rel='stylesheet' href='https://localhost:3000/assets/index-28fadf32.css'></link>
    <meta name='viewport' content='width=device-width, initial-scale=1.0'></meta>
  </head>

  <body>
    <script type='module' src='https://localhost:3000/src/main.js'></script>
    <script type='module'>
      import y from 'y'
import xx from './xx'
document.write(xx)
console.log(y)
    </script>
    <script>
      window.addEventListener('click', async () => {
        await import('./esm.js')
})
    </script>
  </body>
</html>
</template>"
`;
